substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/UniqueShortNameProvider.java

```diff
@@ -0,0 +1,163 @@
+package com.oracle.svm.core;
+
+import com.oracle.svm.core.util.VMError;
+import com.oracle.svm.util.ReflectionUtil;
+import jdk.vm.ci.meta.ResolvedJavaMethod;
+import jdk.vm.ci.meta.ResolvedJavaType;
+import jdk.vm.ci.meta.Signature;
+import org.graalvm.nativeimage.ImageSingletons;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Executable;
+import java.lang.reflect.Field;
+import java.lang.reflect.Member;
+import java.lang.reflect.Method;
+
+/**
+ * An abstract provider class that defines a protocol for generating unique short names for
+ * Java methods and fields which are compatible for use as local linker symbols. An instance
+ * of this class may be registered as an image singleton in order to ensure that a specific
+ * naming convention is adopted.
+ */
+public abstract class UniqueShortNameProvider {
+    private static UniqueShortNameProvider provider = null;
+
+    public static synchronized UniqueShortNameProvider provider() {
+        if (provider == null) {
+            if (ImageSingletons.contains(UniqueShortNameProvider.class)) {
+                provider = ImageSingletons.lookup(UniqueShortNameProvider.class);
+            } else {
+                provider = new DefaultUniqueShortNameProvider();
+                ImageSingletons.add(UniqueShortNameProvider.class, provider);
+            }
+        }
+        return provider;
+    }
+
+    /**
+     * Returns a short, reasonably descriptive, but still unique name for the provided method
+     * which can be used as a linker local symbol.
+     */
+    public abstract String uniqueShortName(ResolvedJavaMethod method);
```

Comment:
I was fairly sure that this was not going to be work because of the recursive invocations of the `uniqueShortname` and `classLoaderId` methods in the BFD provider. However, you are right that the provider can be simplified if we move the wrapper method code out into `SubstrateUtil` and/or the BFD provider.

I have condensed the interface down so that we have the following 3 API methods:

```
static UniqueShortNameProvider singleton();
String uniqueShortName(ClassLoader, ResolvedJavaType, String, Signature, boolean);
String uniqueShortName(Member);
```

The API on `SubstrateUtil` provides the following methods which redirect where needed

```
String uniqueShortName(ResolvedJavaMethod);
String uniqueShortName(ClassLoader, ResolvedJavaType, String, Signature, boolean);
String uniqueShortName(Member m);
String uniqueStubName(ResolvedJavaMethod m);
```

I also restored this method to `SubstrateUtil`
```
static String classLoaderNameAndId(ClassLoader loader);
```

The BFD provider implements its own internal loader name generator, also called `classLoaderNameAndId`, which avoids generating names for builtin and Graal loaders. It piggybacks on the implementation in `SustrateUtil`.